File: app/src/main/java/io/github/zwieback/familyfinance/business/operation/dialog/ExpenseOperationFilterDialog.java
package io.github.zwieback.familyfinance.business.operation.dialog;

import android.os.Bundle;
import android.widget.EditText;
import androidx.annotation.StringRes;

import com.johnpetitto.validator.ValidatingTextInputLayout;

import org.jetbrains.annotations.NotNull;

import io.github.zwieback.familyfinance.R;
import io.github.zwieback.familyfinance.business.article.activity.ExpenseArticleActivity;
import io.github.zwieback.familyfinance.business.operation.filter.ExpenseOperationFilter;
import io.github.zwieback.familyfinance.core.dialog.EntityFilterDialog;
import io.github.zwieback.familyfinance.databinding.DialogFilterExpenseOperationBinding;
import io.github.zwieback.familyfinance.widget.ClearableEditText;

import static io.github.zwieback.familyfinance.business.operation.filter.ExpenseOperationFilter.EXPENSE_OPERATION_FILTER;

public class ExpenseOperationFilterDialog extends OperationWithArticleFilterDialog<
        ExpenseOperationFilter, DialogFilterExpenseOperationBinding, ExpenseArticleActivity> {

    public static ExpenseOperationFilterDialog newInstance(ExpenseOperationFilter filter) {
        ExpenseOperationFilterDialog fragment = new ExpenseOperationFilterDialog();
        Bundle args = Companion.createArguments(EXPENSE_OPERATION_FILTER, filter);
        fragment.setArguments(args);
        return fragment;
    }

    public static ExpenseOperationFilterDialog newInstance(ExpenseOperationFilter filter,
                                                           @StringRes int dialogTitleId) {
        ExpenseOperationFilterDialog fragment = new ExpenseOperationFilterDialog();
        Bundle args = Companion.createArguments(EXPENSE_OPERATION_FILTER, filter);
        args.putInt(EntityFilterDialog.DIALOG_TITLE, dialogTitleId);
        fragment.setArguments(args);
        return fragment;
    }

    @NotNull
    @Override
    protected final ExpenseOperationFilter createCopyOfFilter(ExpenseOperationFilter filter) {
        return new ExpenseOperationFilter(filter);
    }

    @Override
    protected final String getInputFilterName() {
        return EXPENSE_OPERATION_FILTER;
    }

    @Override
    protected final int getDialogTitle() {
        return R.string.expense_operation_filter_title;
    }

    @Override
    protected final int getDialogLayoutId() {
        return R.layout.dialog_filter_expense_operation;
    }

    @Override
    protected final void bind(@NotNull ExpenseOperationFilter filter) {
        getBinding().setFilter(filter);
        super.bind(filter);
    }

    @Override
    final Class<ExpenseArticleActivity> getArticleActivityClass() {
        return ExpenseArticleActivity.class;
    }

    @Override
    final ClearableEditText getArticleEdit() {
        return getBinding().article;
    }

    @Override
    final ClearableEditText getAccountEdit() {
        return getBinding().account;
    }

    @Override
    final ClearableEditText getOwnerEdit() {
        return getBinding().owner;
    }

    @Override
    final ClearableEditText getCurrencyEdit() {
        return getBinding().currency;
    }

    @Override
    final EditText getStartDateEdit() {
        return getBinding().startDate;
    }

    @Override
    final EditText getEndDateEdit() {
        return getBinding().endDate;
    }

    @Override
    final EditText getStartValueEdit() {
        return getBinding().startValue;
    }

    @Override
    final EditText getEndValueEdit() {
        return getBinding().endValue;
    }

    @Override
    final ValidatingTextInputLayout getStartDateLayout() {
        return getBinding().startDateLayout;
    }

    @Override
    final ValidatingTextInputLayout getEndDateLayout() {
        return getBinding().endDateLayout;
    }

    @Override
    final ValidatingTextInputLayout getStartValueLayout() {
        return getBinding().startValueLayout;
    }

    @Override
    final ValidatingTextInputLayout getEndValueLayout() {
        return getBinding().endValueLayout;
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/operation/dialog/ExpenseOperationFilterDialog.kt
package io.github.zwieback.familyfinance.business.operation.dialog

import android.widget.EditText
import androidx.annotation.StringRes
import com.johnpetitto.validator.ValidatingTextInputLayout
import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.business.article.activity.ExpenseArticleActivity
import io.github.zwieback.familyfinance.business.operation.filter.ExpenseOperationFilter
import io.github.zwieback.familyfinance.business.operation.filter.ExpenseOperationFilter.Companion.EXPENSE_OPERATION_FILTER
import io.github.zwieback.familyfinance.databinding.DialogFilterExpenseOperationBinding
import io.github.zwieback.familyfinance.widget.ClearableEditText

class ExpenseOperationFilterDialog :
    OperationWithArticleFilterDialog<ExpenseOperationFilter, DialogFilterExpenseOperationBinding, ExpenseArticleActivity>() {

    override val inputFilterName: String
        get() = EXPENSE_OPERATION_FILTER

    override val dialogTitle: Int
        get() = R.string.expense_operation_filter_title

    override val dialogLayoutId: Int
        get() = R.layout.dialog_filter_expense_operation

    override val articleActivityClass: Class<ExpenseArticleActivity>
        get() = ExpenseArticleActivity::class.java

    override val articleEdit: ClearableEditText
        get() = binding.article

    override val accountEdit: ClearableEditText
        get() = binding.account

    override val ownerEdit: ClearableEditText
        get() = binding.owner

    override val currencyEdit: ClearableEditText
        get() = binding.currency

    override val startDateEdit: EditText
        get() = binding.startDate

    override val endDateEdit: EditText
        get() = binding.endDate

    override val startValueEdit: EditText
        get() = binding.startValue

    override val endValueEdit: EditText
        get() = binding.endValue

    override val startDateLayout: ValidatingTextInputLayout
        get() = binding.startDateLayout

    override val endDateLayout: ValidatingTextInputLayout
        get() = binding.endDateLayout

    override val startValueLayout: ValidatingTextInputLayout
        get() = binding.startValueLayout

    override val endValueLayout: ValidatingTextInputLayout
        get() = binding.endValueLayout

    override fun createCopyOfFilter(filter: ExpenseOperationFilter): ExpenseOperationFilter {
        return ExpenseOperationFilter(filter)
    }

    override fun bind(filter: ExpenseOperationFilter) {
        binding.filter = filter
        super.bind(filter)
    }

    companion object {
        fun newInstance(filter: ExpenseOperationFilter) = ExpenseOperationFilterDialog().apply {
            arguments = createArguments(EXPENSE_OPERATION_FILTER, filter)
        }

        fun newInstance(
            filter: ExpenseOperationFilter,
            @StringRes dialogTitleId: Int
        ): ExpenseOperationFilterDialog {
            return ExpenseOperationFilterDialog().apply {
                arguments = createArguments(EXPENSE_OPERATION_FILTER, filter).apply {
                    putInt(DIALOG_TITLE, dialogTitleId)
                }
            }
        }
    }
}
